crates/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -122,21 +146,115 @@ impl PerBlockContext {
+            dfg.set_value_from_id(result_value, *new_value);
         }
 
         // Delete load instructions
         // Even though we could let DIE handle this, doing it here makes the debug output easier
         // to read.
-        dfg[self.block_id]
+        dfg[block_id]
             .instructions_mut()
             .retain(|instruction| !loads_to_substitute.contains_key(instruction));
 
         protected_allocations
     }
 
+    // This method finds the load values to substitute for a given address
+    // The search starts at the block supplied as a parameter. If there is not a load to substitute
+    // the CFG is analyzed to determine whether a predecessor block has a load value to substitute.
+    // If there is no load value to substitute the original address is returned.
+    fn fetch_load_value_to_substitute(&self, block_id: BasicBlockId, address: ValueId) -> ValueId {
+        let mut stack = vec![block_id];
+        let mut visited = HashSet::new();
+
+        let mut dom_tree = DominatorTree::with_cfg_and_post_order(&self.cfg, &self.post_order);
+        while let Some(block) = stack.pop() {
+            visited.insert(block);
+
+            // Check whether there is a load value to substitute in the current block.
+            // Return the value if found.
+            if let Some((value, load_block_id)) =
+                self.load_values_to_substitute_per_func.get(&address)
+            {
+                if *load_block_id == block {
+                    return *value;
+                }
+            }
+
+            // If no load values to substitute have been found in the current block, check the block's predecessors.
+            let predecessors = self.cfg.predecessors(block);
+            for predecessor in predecessors {
+                if dom_tree.is_reachable(predecessor)
+                    && dom_tree.dominates(predecessor, block)
+                    && !visited.contains(&predecessor)
+                {
+                    stack.push(predecessor);
+                }
+            }
+        }
+        address
+    }
+
+    // This method determines which loads should be substituted.
+    // Starting at the block supplied as a parameter, we check whether a store has occurred with the given address.
+    // If no store has occurred in the supplied block, the CFG is analyzed to determine whether
+    // a predecessor block has a store at the given address.
+    fn find_load_to_substitute(
+        &mut self,
+        block_id: BasicBlockId,
+        address: ValueId,
+        dfg: &DataFlowGraph,
+        instruction_id: &InstructionId,
+        loads_to_substitute: &mut BTreeMap<InstructionId, ValueId>,
+        load_values_to_substitute_per_block: &mut BTreeMap<ValueId, ValueId>,
+    ) -> bool {
+        let mut stack = vec![block_id];
+        let mut visited = HashSet::new();
+
+        let mut dom_tree = DominatorTree::with_cfg_and_post_order(&self.cfg, &self.post_order);
```

Comment:
I will submit an update in a moment to move this out as the dom tree is being recomputed per call to find_load_substitute which is inefficient.

Reply:
Fixed now